src/avram/attribute.cr

```diff
@@ -1,77 +1,69 @@
 class Avram::Attribute(T)
   alias ErrorMessage = String | Proc(String, String, String) | Avram::CallableErrorMessage
-  getter :name
-  setter :value
-  getter :param_key
+  getter name : Symbol
+  setter value : T?
+  getter param_key : String
   @errors = [] of String
+  @param : Avram::Uploadable | String?
 
-  def initialize(
-    @name : Symbol,
-    @param : Avram::Uploadable | String?,
-    @value : T,
-    @param_key : String
-  )
+  def initialize(@name, @param, @value : T?, @param_key)
     @original_value = @value
   end
 
   @_permitted : Avram::PermittedAttribute(T)?
 
   def permitted
     @_permitted ||= begin
-      Avram::PermittedAttribute.new(name: @name, param: @param, value: @value, param_key: @param_key).tap do |attribute|
+      Avram::PermittedAttribute(T).new(name: @name, param: @param, value: @value, param_key: @param_key).tap do |attribute|
         errors.each do |error|
           attribute.add_error error
         end
       end
     end
   end
 
-  def param
+  def param : Avram::Uploadable | String
     @param || value.to_s
   end
 
-  def add_error(message : ErrorMessage = "is invalid")
-    perform_add_error(message)
-  end
-
-  private def perform_add_error(message : String = "is invalid")
+  def add_error(message : String = "is invalid")
     @errors << message
   end
 
-  private def perform_add_error(message : (Proc | Avram::CallableErrorMessage))
+  def add_error(message : (Proc | Avram::CallableErrorMessage))
     message_string = message.call(@name.to_s, @value.to_s)
-    perform_add_error(message_string)
+    add_error(message_string)
   end
 
   def reset_errors
     @errors = [] of String
   end
 
-  def errors
+  def errors : Array(String)
     @errors.uniq
   end
 
-  def value
+  def value : T?
     ensure_no_blank(@value)
   end
 
-  def original_value
+  def original_value : T?
     ensure_no_blank(@original_value)
   end
 
-  private def ensure_no_blank(value : T)
+  private def ensure_no_blank(value : T?) : T?
     if value.is_a?(Avram::Uploadable | String) && value.blank?
       nil
     else
       value
```

Comment:
I wonder if we should be adding a `not_nil!` here so the compiler knows this type is locked in? 🤔

Reply:
The value can always be nil, this code is just the logic that also treats blank strings as nil.